lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
```

Comment:
1. The name is misleading I think. More like `OncePerThreadMutex`?
  The concept of recursive mutex is non-straightforward and negating it makes it even less obvious.
2. I think what we're looking for is a regular `scoped_lock` with extra assertions, so instead of creating a new "mutex" class I'd create a new "lock" class (`OncePerThreadScopedLock`?) which embeds a regular `scoped_lock` and has extra checks in the constructor/destructor.

Reply:
1. Ok on the name
2. It can be done like that, but then you'd need an extra `static` data structure somewhere that keeps tabs on the (one or many) `mutex -> thead-id` relations